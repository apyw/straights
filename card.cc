#include "card.h"
#include <vector>

Card::Card(Suit s, Rank r) : suit{s}, rank{r} {}

Suit Card::getSuit() const{
    return suit;
}

Rank Card::getRank() const{
    return rank;
}

bool operator==(const Card &a, const Card &b) {
    return a.getSuit() == b.getSuit() && a.getRank() == b.getRank();
}

std::ostream &operator<<(std::ostream &out, const Card &card) {
    std::vector<char> suits = { 'C', 'D', 'H', 'S' };
	std::vector<char> ranks = { 'A', '2', '3', '4', '5', '6',
		'7', '8', '9', 'T', 'J', 'Q', 'K' };
    out << ranks[card.getRank()] << suits[card.getSuit()];
    return out;
}
