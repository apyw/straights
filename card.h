/*
 * Straights card game implementation
 *
 * Card class.  Is responsible for keeping track of value of card.  
 *
 */

#ifndef CARD_H
#define CARD_H

#include <iostream>

enum Suit { club, diamond, heart, spade };
enum Rank { ace, two, three, four, five, six, seven, eight, 
            nine, ten, jack, queen, king};

class Card {
    Suit suit;
    Rank rank;
  public:
    Card(Suit, Rank);
    Suit getSuit() const;
    Rank getRank() const;
};

bool operator==(const Card &a, const Card &b);
std::ostream &operator<<(std::ostream &out, const Card &card);

#endif 
