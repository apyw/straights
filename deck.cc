#include "deck.h"
#include <algorithm>

const int suitNum = 4;
const int rankNum = 13;
const int deckSize = 52; 

Deck::Deck() {
    for (int i = 0; i < suitNum; i++) {
        for (int j = 0; j < rankNum; j++) {
            cardDeck.push_back(new Card(static_cast<Suit>(i), static_cast<Rank>(j)));
        }
    }
}

Deck::~Deck() {
    for(unsigned int i = 0; i < cardDeck.size(); i++) {
        delete cardDeck[i];
    }
    cardDeck.clear();
}

std::vector<Card*> Deck::getCardDeck() const{
    return cardDeck;
}

// source: codeForStudents/shuffle.cc
void Deck::shuffle(std::default_random_engine rng) {
    std::shuffle(cardDeck.begin(), cardDeck.end(), rng );
}

std::vector<Card*> Deck::deal(int playerID) {
    std::vector<Card*> hand;
    for (int i = 0; i < 13; i++) {
        hand.push_back(cardDeck[playerID * 13 + i]);
    }
    return hand;
}

std::ostream &operator<<(std::ostream &out, const Deck &deck) {
    for (int i = 0; i < deckSize; i++) { 
        out << *(deck.getCardDeck()[i]) << " ";
        if ((i+1)%13 == 0) out << std::endl;
    }
    return out;
}
