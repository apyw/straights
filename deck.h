/*
 * Straights card game implementation
 *
 * Deck class.  Is responsible for keeping track of order of deck.  
 *
 */

#ifndef DECK_H
#define DECK_H

#include "card.h"
#include <vector>
#include <random>

class Deck {
    std::vector<Card*> cardDeck;
  public:
    Deck(); 
    ~Deck();
    std::vector<Card*> getCardDeck() const;
    void shuffle(std::default_random_engine rng);
    std::vector<Card*> deal(int playerID);
};

std::ostream &operator<<(std::ostream &out, const Deck &deck);

#endif
