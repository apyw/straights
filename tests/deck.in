h
c
c
c
deck
