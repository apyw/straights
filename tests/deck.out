Is Player1 a human (h) or a computer (c)?
>Is Player2 a human (h) or a computer (c)?
>Is Player3 a human (h) or a computer (c)?
>Is Player4 a human (h) or a computer (c)?
>A new round begins. It's Player4's turn to play.
Player4 plays 7S.
Cards on the table:
Clubs:
Diamonds:
Hearts:
Spades: 7
Your hand: 9H 3H KH 7C 3S 8H 5C QH 8D TH 2D TD 3C
Legal plays: 7C
>9H 3H KH 7C 3S 8H 5C QH 8D TH 2D TD 3C
6H JH 7D 5H 2S 4C JC KD 8C 7H JS AS QD
QS KC 9D 4S 4H 6C JD AC AH 9S 2C TC 3D
AD 6D TS 9C 8S 2H KS 4D 5D QC 7S 6S 5S
>